tools/util/OperationManager.hpp

```diff
@@ -0,0 +1,1172 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2016-19, Lawrence Livermore National Security, LLC and Umpire
+// project contributors. See the COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (MIT)
+//////////////////////////////////////////////////////////////////////////////
+#ifndef REPLAY_OperationManager_HPP
+#define REPLAY_OperationManager_HPP
+
+#include <iostream>
+#include <cstdint>
+#include <vector>
+
+#include "umpire/Allocator.hpp"
+#include "umpire/strategy/AllocationAdvisor.hpp"
+#include "umpire/strategy/SizeLimiter.hpp"
+#include "umpire/strategy/MixedPool.hpp"
+#include "umpire/strategy/MonotonicAllocationStrategy.hpp"
+#include "umpire/strategy/SlotPool.hpp"
+#include "umpire/strategy/ThreadSafeAllocator.hpp"
+#include "umpire/ResourceManager.hpp"
+
+class Operation {
+public:
+ using AllocationOpMap = std::unordered_map<uint64_t, Operation*>;
+  std::function<void ()> op;
+
+  Operation(
+      std::vector<umpire::Allocator*>& alloc_array,
+      AllocationOpMap& alloc_operations
+  ) :   m_alloc_array(alloc_array)
+      , m_alloc_operations(alloc_operations)
+  {
+  }
+
+  void run()
+  {
+    op();
+  }
+
+  //
+  // AllocationAdvisor
+  //
+  template <typename... Args>
+  void bld_advisor(
+      const bool introspection,
+      const std::string& allocator_name,
+      const std::string& base_allocator_name,
+      Args&&... args
+  )
+  {
+    if (introspection) {
+      op = [=]() {
+        auto& rm = umpire::ResourceManager::getInstance();
+
+        rm.makeAllocator<umpire::strategy::AllocationAdvisor, true>
+          ( allocator_name, rm.getAllocator(base_allocator_name),
+            std::forward<Args>(args)...);
+
+        auto allocator = new umpire::Allocator(rm.getAllocator(allocator_name));
+        this->m_alloc_array.push_back(allocator);
+      };
+    }
+    else {
+      op = [=]() {
+        auto& rm = umpire::ResourceManager::getInstance();
+
+        rm.makeAllocator<umpire::strategy::AllocationAdvisor, false>
+          ( allocator_name, rm.getAllocator(base_allocator_name),
+            std::forward<Args>(args)...);
+
+        auto allocator = new umpire::Allocator(rm.getAllocator(allocator_name));
+        this->m_alloc_array.push_back(allocator);
+      };
+    }
+  }
+
+  template <typename... Args>
+  void bld_advisor(
```

Comment:
Same thing here. Use the method above and pass the arguments as they appear below if you need this version.